### PR TITLE
Cert award from template

### LINKED ISSUE #1965: [6.0] When awarding certificate - option to pull in certificate template
**Is your feature request related to a problem? Please describe.**
When awarding a certificate, we want to pull in a template to associate that individual awarded certificate with, so in future updates of the certificate template, it can sync and make it faster to recreate certificate from a course program for example.

https://www.loom.com/share/e6b4c3d2cb3d4ebebac72b1d3ff5f098

**Describe the solution you'd like**
An option on the award certificate edit page to create from certificate template

**Describe alternatives you've considered**
Global blocks (but that only solves the problem of initially creating it rather than updating it from the template down the line)

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

Requires #2026 (submitted as a PR against taht branch for now to reduce codesize for review)
Requires https://github.com/gocodebox/lifterlms-blocks/pull/158

Closes #1965

Adds a button to:

1) Certificate Template block editor sidebar
2) Awarded certificate post table (replaces the default "Add New Button" provided via PHP by the WP core
3) LLMS Student Certificates reporting screen

When pressed, the button pops a modal allowing student and / or template selection based on the context

In 1): student can be selected, template cannot (template is the current template being edited) and the "start from scratch" button is disabled

In 2): both the template and the student can be selected. Start from scratch exists and does not pre-select a student

In 3) template can be selected, not student (student is the current student being veiwed). Start from scratch prefills the current student using query string vars

The only thing I'm uncertain about is the way we change the UI on the post table screen. The button looks *similar* but not identical. Had I thought about this before I started I would have separated this into multiple components so the modal could be popped independently based on the pressing of an existing dom button... It's not too late to do this but I wanted feedback before reworking too much.

## How has this been tested?

+ Manually
+ New tests (php and js unit)
+ I will be writing additional e2e tests tomorrow

## Types of changes
New feature

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: includes/controllers/class-llms-controller-awards.php [hunk 1]
```diff
@@ -106,6 +108,42 @@ public static function on_earn( $user_id, $post_id ) {
 
 	}
 
+	/**
+	 * Awarded certificate REST API insertion callback.
+	 *
+	 * Automatically syncs an awarded certificate with it's parent when inserted via the REST API and
```
Comment (eri-trabiccolo) [inline includes/controllers/class-llms-controller-awards.php]:
```suggestion
	 * Automatically syncs an awarded certificate with its parent when inserted via the REST API and
```

File: src/js/util/award-certificate-button/index.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Button and modal interface for creating a draft certificate from a specified template for a specific student.
+ *
+ * @since [version]
+ *
+ * @param {Object}   params                Component configuration object.
+ * @param {string}   params.modalTitle     [description]
+ * @param {string}   params.buttonLabel    [description]
+ * @param {Boolean}  params.isDisabled     [description]
```
Comment (eri-trabiccolo) [inline src/js/util/award-certificate-button/index.js]: missing real description.

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
This looks like a badass PR <3
Although a couple of things:
1) From the Awarded Certificates post lists, clicking on Award Certificate I can only start from scratch or fill both user + template. If I only fill the user and then click on Start from scratch, the award has not student pre-selected. This behavior is a bit different from when I do the same from the student reporting screen, where "starting from scratch" still carries on the student.
2) Looks like the certificate search doesn't work properly. E.g.
If I have a certificate template with title "Foo Bar", and i search for "foo" it finds the certificate "Foo Bar", but if I type "foo bar" (or "Foo Bar") the certificate above is not found. I proposed a fix in #2028

**Comment by thomasplevy** [comment]
whoops didn't intend to merge

File: src/js/util/award-certificate-button/index.js (hunk 2 above)
Comment (thomasplevy) [reply src/js/util/award-certificate-button/index.js]: https://github.com/gocodebox/lifterlms/commit/b878a01cc6c969528d768f1dc79a16b516a4cd4c

File: includes/controllers/class-llms-controller-awards.php (hunk 1 above)
Comment (thomasplevy) [reply includes/controllers/class-llms-controller-awards.php]: https://github.com/gocodebox/lifterlms/commit/b878a01cc6c969528d768f1dc79a16b516a4cd4c

**Comment by thomasplevy** [comment]
> 1. From the Awarded Certificates post lists, clicking on Award Certificate I can only start from scratch or fill both user + template. If I only fill the user and then click on Start from scratch, the award has not student pre-selected. This behavior is a bit different from when I do the same from the student reporting screen, where "starting from scratch" still carries on the student.

https://github.com/gocodebox/lifterlms/commit/fcd954841c966684822fee42912ad6a96667d22b

**Pull request merged**